### ISSUE TITLE
[SPIR-V] Unexpected label emitted in SPIRVAsmPrinter when debug info is present

### ISSUE THREAD

**Issue by seven-mile** [llvm/llvm-project#102732]
I used `-g` to emit debug info in Clang and get a LLVM IR module. When I tried to invoke `llc` and use SPIR-V backend to emit the final SPIR-V, an unexpected label is emitted after `OpFunctionEnd`:

```spirv
... omitted
        OpBranchConditional %56 %61 %59
        %59 = OpLabel
        OpReturn
        OpFunctionEnd
Lfunc_end0:
                                        ; -- End function
```
, which prevented the SPIR-V to be consumed by `spirv-as`. After disabling `-g`, the label is gone. I believe it's caused by these lines in the base class `AsmPrinter`:

https://github.com/llvm/llvm-project/blob/9a227ba3e129eaa89498b97421afefb1e9d681df/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp#L1945-L1949

We need some overrides to disable this behaviour for SPIR-V.

**Comment by VyacheslavLevytskyy** [comment]
@seven-mile Thank you for the report! I don't see any good workarounds using existing hooks/overrides. `needFuncLabels()` is always true if we have debug info, and a flag for the debug info is private (https://github.com/llvm/llvm-project/blob/9a227ba3e129eaa89498b97421afefb1e9d681df/llvm/include/llvm/CodeGen/AsmPrinter.h#L229).

A path that I see is that we fix it as a part of AsmPrinter.cpp itself by checking `!TT.isSPIRV()` before emitting a label after the function body and referring to the SPIR-V specification that supports label instructions only inside a block, not after the function body.

**Comment by seven-mile** [comment]
Resolved by #107013
